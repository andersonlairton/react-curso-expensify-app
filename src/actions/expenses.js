import uuid from 'uuid';

export const addExpense = (
    {
         descricao = '',
          nota = '', 
          amount = 0, 
          createAT = 0
         } = {}) => (
    {
        type: 'ADD_EXPENSE',
        expense:
        {
            id: uuid(),
            descricao,
            nota,
            amount,
            createAT
        }
    });
//função que remove
export const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

//função que edita

export const editExpense = (id, updates) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
);