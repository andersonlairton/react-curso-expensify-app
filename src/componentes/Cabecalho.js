import React from 'react';
import {BrowserRouter,Route,Switch,Link,NavLink} from 'react-router-dom';

const Cabecalho=()=>
(
    <div>
        <h1>Exepensivo App</h1>
        <NavLink to="/" activeClassName="is-active"/* chamando elementos da classe */ exact={true}>Home</NavLink>
        <NavLink to="/help" activeClassName="is-active">Ajuda</NavLink>
        <NavLink to="/edit" activeClassName="is-active">Editar</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create page</NavLink>
    </div>
);

export default Cabecalho;