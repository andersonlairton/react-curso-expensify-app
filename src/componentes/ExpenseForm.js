import React from 'react'
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dateS';
 //import 'react-dates/lib/css/_datepicker.css';

//const now = moment();//CONFIGURANDO DATA
//console.log(now.format('MMM Do,YYYY'));
export default class ExpenseForm extends React.Component {
    constructor (props)
    {
        super(props);
        this.state =
        {
            descricao: props.expense?props.expense.descricao:'',
            note: props.expense?props.expense.note:'',
            amount: props.expense?(props.expense.amount/100).toString():'',
            createAt: props.expense? moment(props.expense.createAt):moment(),
            calendarFocused: false,
            error:''
        };
    }
    
    onDescriptionChange = (e) => {
        const descricao = e.target.value;
        this.setState(() => ({ descricao }));
    }
    onNoteCharge = (n) => {
        const note = n.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (a) => {
        const amount = a.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createAt) => {
        if (createAt) {
            this.setState(() => ({ createAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }
/* 
    enviarForm=(e)=> {
        e.preventDefault();
        if(!this.state.descricao||!this.state.amount)
        {
            this.setState(()=>({error:'digite os dados para enviar'}));
            console.log('digite dados para enviar');

        }
        else
        {
            this.setState(()=>({error:''}));
           // console.log('enviado');
            //enviado dados
            this.props.onSubmit(
                {
                    descricao:this.state.descricao,
                    amount:parseFloat(this.state.amount,10)*100,
                    createAt:this.state.createAt.valueOf(),
                    note:this.setState.note
                });
        }

    }*/
   enviarForm = (e) => {
        e.preventDefault();
    
        if (!this.state.descricao || !this.state.amount) 
        {
          this.setState(() => ({ error: 'digite os dados para enviar' }));
        } else 
        {
          this.setState(() => ({ error: '' }));
          this.props.onSubmit(
            {
            descricao: this.state.descricao,
            amount: parseFloat(this.state.amount, 10) * 100,
            createAt:this.state.createAt.valueOf(),
            note: this.state.note
            }
            );
        }
      };

      
    render() {
        return (
            <div>
            {this.state.error &&<p>{this.state.error}</p>}
                <fieldset>
                    Adicionar despensa
                </fieldset>
                <form onSubmit={this.enviarForm}>
                    <input
                        type="text"
                        placeholder="descrição"
                        autoFocus
                        value={this.state.descricao}
                        onChange={this.onDescriptionChange}
                    />

                    <input type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}

                    />

                    <SingleDatePicker//configurando campo data
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="adicionar nota é opcional"
                        value={this.state.note}
                        onChange={this.onNoteCharge}
                    />
                    <button>Adicionar despensa</button>
                </form>
            </div>
        )
    }
}
/*
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
        descricao: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        descricao: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}*/
