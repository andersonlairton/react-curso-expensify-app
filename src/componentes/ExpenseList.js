import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) =>
    (
        <div>
            <h1>Componente da pagina ExpenseList</h1>

            {props.expenses.length === 0 ? (
                <p>no expenses</p>
            ) : (


                    //  {<p>filtro:{props.filters.text}</p> }

                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />;
                    }
                    )
                    )
            }




        </div>
    );
const mapStateToProps = (state) => {
    return {
        //filtrando dados conforme filtros da pagina selectExpenses
        expenses: selectExpenses(state.expenses, state.filters),
        //  filters:state.filters
        /*
        expenses: state.expenses, //pegando dados da loja 
        filters:state.filters
        */
    };
}
export default connect(mapStateToProps)(ExpenseList);
