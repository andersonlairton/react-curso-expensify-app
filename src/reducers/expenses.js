const expenseReducerDefaultState=[];

export default (state =expenseReducerDefaultState,action)=>
{
    switch (action.type) 
    {
        case 'ADD_EXPENSE':
            return[...state,
                action.expense
            ];//retorne os dados da matriz
        case 'REMOVE_EXPENSE':
            return state.filter(({id})=> id !==action.id);//se o id for igual ao action.id resultara em falso
        case 'EDIT_EXPENSE':
            return state.map((expense)=>
            {
                if(expense.id===action.id)
                {
                    return{
                        ...expense,
                        ...action.updates
                    }
                }
                else
                {
                    return expense;
                }
            });
        default:
            return state;
    }
};

