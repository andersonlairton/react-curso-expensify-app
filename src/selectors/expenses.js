/*import moment from 'moment';
//classificando dados
export default (expenses,/*criando variaveis e filtros individuais { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createATMoment = moment(expense.createAT);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createATMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createATMoment, 'day') : true;
        // const startDateMatch= typeof startDate !=='number' || expense.createAT>=startDate;
        //const endDateMatch= typeof endDate !=='number'|| expense.createAT<=endDate;
        const textMatch = expense.descricao.toLowerCase().includes(text.toLowerCase());//pesquisando dados e convertendo eles para letras maiusculas

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) =>//classificando 
    {
        if (sortBy === 'date') {
            return a.createAT < b.createAT ? 1 : -1;
        } else if (sortBy === 'amount')//classificando as despesas por valores
        {
            return a.amount < b.amount ? 1 : -1;
        }
    });
}
*/
import moment from 'moment';

// Get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.descricao.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};
