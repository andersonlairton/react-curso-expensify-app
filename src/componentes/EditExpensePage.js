import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

/*const EditExpensePage=(props)=>{

    console.log(props);
    return(
    <div>
    pagina de edição do id {props.match.params.id}
    </div>)
};*/

export class EditExpensePage extends React.Component{//definindo elemento de classe
  
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);//editanado
    this.props.history.push('/');//redirecionando
  }
  onRemove = () => {//removendo itens
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  render() {
    return(
    <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove}>Remove</button>
    </div>
    );
  };
}

const mapStateToProps = (state, props) => (
  {
  
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
);


const mapDispachToProps = (dispatch,props) => (
  {
    editExpense:(id,expense)=>dispatch(editExpense(id, expense)),
    removeExpense:(data)=>dispatch(removeExpense(data))

  }
);

export default connect(mapStateToProps,mapDispachToProps)(EditExpensePage);