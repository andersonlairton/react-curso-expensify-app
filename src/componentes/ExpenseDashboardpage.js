import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
const ExpenseDashboardpage=()=>(
    <div>
        este Ã© meu componente dashboard
        <ExpenseListFilter/>
        <ExpenseList/>
    </div>
);
export default ExpenseDashboardpage