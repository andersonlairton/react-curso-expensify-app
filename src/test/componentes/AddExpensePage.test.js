import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../componentes/AddExpenseDashboardpage';
import expenses from '../fixuters/expenses';

let addExpense, history, wrapper //definindo variaveis

beforeEach(() => {
    //incrementando suas funçõe e dados
    addExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />);

});

test('addexpense page 2', () => {

    expect(wrapper).toMatchSnapshot();
});

test('onSubmit', () => {

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenCalledWith('/');
    expect(addExpense).toHaveBeenCalledWith(expenses[1]);
});