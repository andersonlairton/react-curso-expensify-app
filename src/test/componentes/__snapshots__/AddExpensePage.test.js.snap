// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addexpense page 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddExpensePage
    history={
        Object {
            "push": [Function],
          }
    }
    onSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          formulario para adicionar despesas
</h1>,
        <ExpenseForm
          onSubmit={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "formulario para adicionar despesas",
        },
        "ref": null,
        "rendered": "formulario para adicionar despesas",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            formulario para adicionar despesas
</h1>,
          <ExpenseForm
            onSubmit={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "formulario para adicionar despesas",
          },
          "ref": null,
          "rendered": "formulario para adicionar despesas",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`addexpense page 2 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddExpensePage
    addExpense={[Function]}
    history={
        Object {
            "push": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          formulario para adicionar despesas
</h1>,
        <ExpenseForm
          onSubmit={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "formulario para adicionar despesas",
        },
        "ref": null,
        "rendered": "formulario para adicionar despesas",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            formulario para adicionar despesas
</h1>,
          <ExpenseForm
            onSubmit={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "formulario para adicionar despesas",
          },
          "ref": null,
          "rendered": "formulario para adicionar despesas",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
